system/Cache/Handlers/RedisHandler.php

```diff
@@ -61,113 +66,166 @@ public function __destruct()
+                $this->connectToRedisServer($config);
+            }
+        } catch (Throwable $t) {
+            // Either connection function can throw Exception, RedisException, or RedisClusterException. Catch any
+            // possible error and log it.
+            $redisMode = $isCluster ? 'Redis' : 'RedisCluster';
+            $message   = "Cache: {$redisMode}: {$t->getMessage()}";
+            log_message('error', $message);
+
+            throw new CriticalError($message, 0, $t);
         }
 
-        switch ($data['__ci_type']) {
-            case 'array':
-            case 'object':
-                return unserialize($data['__ci_value']);
-
-            case 'boolean':
-            case 'integer':
-            case 'double': // Yes, 'double' is returned and NOT 'float'
-            case 'string':
-            case 'NULL':
-                return settype($data['__ci_value'], $data['__ci_type']) ? $data['__ci_value'] : null;
-
-            case 'resource':
-            default:
-                return null;
+        // Use the php serializer to handle everything we set/get from redis.
+        $this->redis->setOption(Redis::OPT_SERIALIZER, Redis::SERIALIZER_PHP);
+
+        if (! empty($this->prefix)) {
+            // Let the driver handle the prefix.
+            $this->redis->setOption(Redis::OPT_PREFIX, $this->prefix);
         }
     }
 
     /**
      * {@inheritDoc}
      */
-    public function save(string $key, $value, int $ttl = 60)
+    public function get(string $key)
     {
-        $key = static::validateKey($key, $this->prefix);
-
-        switch ($dataType = gettype($value)) {
-            case 'array':
-            case 'object':
-                $value = serialize($value);
-                break;
-
-            case 'boolean':
-            case 'integer':
-            case 'double': // Yes, 'double' is returned and NOT 'float'
-            case 'string':
-            case 'NULL':
-                break;
-
-            case 'resource':
-            default:
-                return false;
-        }
+        $key  = static::validateKey($key);
+        $data = $this->redis->get($key);
 
-        if (! $this->redis->hMSet($key, ['__ci_type' => $dataType, '__ci_value' => $value])) {
-            return false;
-        }
+        return $data === false ? null : $data;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function save(string $key, $value, int $ttl = 60)
+    {
+        $key = static::validateKey($key);
 
-        if ($ttl) {
-            $this->redis->expireAt($key, time() + $ttl);
+        if ($ttl > 0) {
+            $rtnVal = $this->redis->setEx($key, $ttl, $value);
+        } else {
+            $rtnVal = $this->redis->set($key, $value);
         }
 
-        return true;
+        return $rtnVal;
     }
 
     /**
      * {@inheritDoc}
      */
     public function delete(string $key)
     {
-        $key = static::validateKey($key, $this->prefix);
+        $key = static::validateKey($key);
```

Comment:
We set the prefix via `setOption` above at line 192, so no need to continue passing it to `validateKey`.

Redis allows a key that is 512 mb in length